oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/introspection/OAuth2TokenIntrospectionClaimAccessor.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.core.introspection;
+
+import java.net.URL;
+import java.time.Instant;
+import java.util.List;
+
+import org.springframework.lang.Nullable;
+import org.springframework.security.oauth2.core.ClaimAccessor;
+
+/**
+ * A {@link ClaimAccessor} for the &quot;claims&quot; that may be contained in the
+ * Introspection Response.
+ *
+ * @author David Kovac
+ * @since 5.6
+ * @see ClaimAccessor
+ * @see OAuth2TokenIntrospectionClaimNames
+ * @see <a target="_blank" href=
+ * "https://tools.ietf.org/html/rfc7662#section-2.2">Introspection Response</a>
+ */
+public interface OAuth2TokenIntrospectionClaimAccessor extends ClaimAccessor {
+
+	/**
+	 * Returns the indicator {@code (active)} whether or not the token is currently active
+	 * @return the indicator whether or not the token is currently active
+	 */
+	default boolean isActive() {
+		return Boolean.TRUE.equals(this.getClaimAsBoolean(OAuth2TokenIntrospectionClaimNames.ACTIVE));
+	}
+
+	/**
+	 * Returns a human-readable identifier {@code (username)} for the resource owner that
+	 * authorized the token
+	 * @return a human-readable identifier for the resource owner that authorized the
+	 * token
+	 */
+	@Nullable
+	default String getUsername() {
+		return this.getClaimAsString(OAuth2TokenIntrospectionClaimNames.USERNAME);
+	}
+
+	/**
+	 * Returns the client identifier {@code (client_id)} for the token
+	 * @return the client identifier for the token
+	 */
+	@Nullable
+	default String getClientId() {
+		return this.getClaimAsString(OAuth2TokenIntrospectionClaimNames.CLIENT_ID);
+	}
+
+	/*
+	 * NOTE: Intention is to change return type of getScope() from String to List<String>
+	 * and then deprecate getScopes() method.
+	 */
+
+	/**
+	 * Returns the scopes {@code (scope)} associated with the token
+	 * @return the scopes associated with the token
+	 * @deprecated Since 5.6. Use {@link #getScopes()} instead
```

Comment:
I don't think we should have both `getScope()` and `getScopes()` as it may be confusing. The preference would be to have only the following signature `List<String> getScope()`

Reply:
First, I agree we should have signature `List<String> getScope()`.

But if we want to preserve backward compatibility (with 5.4.x and 5.5.x versions), `OAuth2IntrospectionAuthenticatedPrincipal` should still implement deprecated `OAuth2IntrospectionClaimAccessor`.
That means there still must be signature `String getScope()`, which in turn will clash with signature `List<String> getScope()`.

So for now it is better to move `String getScope()` to `OAuth2IntrospectionClaimAccessor` and introduce  `List<String> getScope*s*()` in `OAuth2TokenIntrospectionClaimAccessor`.

What do you think?

Reply:
Sorry @qavid, I actually meant to say `List<String> getScopes()` (plural). So I agree that we should go with:

> So for now it is better to move `String getScope()` to `OAuth2IntrospectionClaimAccessor` and introduce  `List<String> getScope*s*()` in `OAuth2TokenIntrospectionClaimAccessor`.